beacon-chain/core/altair/epoch_precompute.go

```diff
@@ -0,0 +1,155 @@
+package altair
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/epoch/precompute"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	"github.com/prysmaticlabs/prysm/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"go.opencensus.io/trace"
+)
+
+// InitializeEpochValidators gets called at the beginning of process epoch cycle to return
+// pre computed instances of validators attesting records and total
+// balances attested in an epoch.
+func InitializeEpochValidators(ctx context.Context, st state.BeaconStateAltair) ([]*precompute.Validator, *precompute.Balance, error) {
+	ctx, span := trace.StartSpan(ctx, "altair.InitializeEpochValidators")
+	defer span.End()
+	pValidators := make([]*precompute.Validator, st.NumValidators())
+	bal := &precompute.Balance{}
+	prevEpoch := helpers.PrevEpoch(st)
+
+	inactivityScores, err := st.InactivityScores()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// This shouldn't happen with a correct beacon state,
+	// but rather be safe to defend against index out of bound panics.
+	if st.NumValidators() > len(inactivityScores) {
+		return nil, nil, errors.New("num of validators can't be greater than length of inactivity scores")
+	}
+	if err := st.ReadFromEveryValidator(func(idx int, val state.ReadOnlyValidator) error {
+		// Was validator withdrawable or slashed
+		withdrawable := prevEpoch+1 >= val.WithdrawableEpoch()
+		pVal := &precompute.Validator{
+			IsSlashed:                    val.Slashed(),
+			IsWithdrawableCurrentEpoch:   withdrawable,
+			CurrentEpochEffectiveBalance: val.EffectiveBalance(),
+			InactivityScore:              inactivityScores[idx],
+		}
+		// Validator active current epoch
+		if helpers.IsActiveValidatorUsingTrie(val, helpers.CurrentEpoch(st)) {
+			pVal.IsActiveCurrentEpoch = true
+			bal.ActiveCurrentEpoch += val.EffectiveBalance()
+		}
+		// Validator active previous epoch
+		if helpers.IsActiveValidatorUsingTrie(val, prevEpoch) {
+			pVal.IsActivePrevEpoch = true
+			bal.ActivePrevEpoch += val.EffectiveBalance()
+		}
+
+		pValidators[idx] = pVal
+		return nil
+	}); err != nil {
+		return nil, nil, errors.Wrap(err, "could not initialize epoch validator")
+	}
+	return pValidators, bal, nil
+}
+
+// ProcessInactivityScores of beacon chain. This updates inactivity scores of beacon chain and
+// updates the precompute validator struct for later processing.
```

Comment:
Can you add some details to how inactivity scores work? Under what conditions should `v.InactivityScore` increase?